File: yasmin_ros/src/yasmin_ros/get_parameters_state.cpp
// Copyright (C) 2025 Miguel Ángel González Santamarta
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include "rclcpp/rclcpp.hpp"
#include "yasmin_ros/basic_outcomes.hpp"

#include "yasmin_ros/get_parameters_state.hpp"

using namespace yasmin_ros;

GetParametersState::GetParametersState(
    const std::map<std::string, std::any> &parameters,
    rclcpp::Node::SharedPtr node)
    : yasmin::State({basic_outcomes::SUCCEED, basic_outcomes::ABORT}),
      parameters_(parameters) {

  if (node == nullptr) {
    this->node_ = YasminNode::get_instance();
  } else {
    this->node_ = node;
  }
}

std::string GetParametersState::execute(
    std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
  for (const auto &param : this->parameters_) {
    const std::string &param_name = param.first;
    const std::any &default_value = param.second;

    if (!this->node_->has_parameter(param_name)) {

      if (default_value.type() == typeid(bool)) {
        this->node_->declare_parameter(param_name,
                                       std::any_cast<bool>(default_value));
      } else if (default_value.type() == typeid(int)) {
        this->node_->declare_parameter(param_name,
                                       std::any_cast<int>(default_value));
      } else if (default_value.type() == typeid(double)) {
        this->node_->declare_parameter(param_name,
                                       std::any_cast<double>(default_value));
      } else if (default_value.type() == typeid(std::string)) {
        this->node_->declare_parameter(
            param_name, std::any_cast<std::string>(default_value));
      } else if (default_value.type() == typeid(std::vector<bool>)) {
        this->node_->declare_parameter(
            param_name, std::any_cast<std::vector<bool>>(default_value));
      } else if (default_value.type() == typeid(std::vector<int>)) {
        this->node_->declare_parameter(
            param_name, std::any_cast<std::vector<int64_t>>(default_value));
      } else if (default_value.type() == typeid(std::vector<double>)) {
        this->node_->declare_parameter(
            param_name, std::any_cast<std::vector<double>>(default_value));
      } else if (default_value.type() == typeid(std::vector<std::string>)) {
        this->node_->declare_parameter(
            param_name, std::any_cast<std::vector<std::string>>(default_value));
      } else if (default_value.type() == typeid(std::vector<uint8_t>)) {
        this->node_->declare_parameter(
            param_name, std::any_cast<std::vector<uint8_t>>(default_value));
      } else {
        YASMIN_LOG_ERROR("Unsupported default type for parameter: '%s'",
                         param_name.c_str());
        return basic_outcomes::ABORT;
      }
    }

    YASMIN_LOG_INFO("Retrieving parameter '%s'", param_name.c_str());

    auto parameter = this->node_->get_parameter(param_name);
    auto type = this->node_->get_parameter_types({param_name});

    switch (type[0]) {
    case rclcpp::ParameterType::PARAMETER_BOOL:
      blackboard->set<bool>(param_name, parameter.as_bool());
      break;
    case rclcpp::ParameterType::PARAMETER_INTEGER:
      blackboard->set<int64_t>(param_name, parameter.as_int());
      break;
    case rclcpp::ParameterType::PARAMETER_DOUBLE:
      blackboard->set<double>(param_name, parameter.as_double());
      break;
    case rclcpp::ParameterType::PARAMETER_STRING:
      blackboard->set<std::string>(param_name, parameter.as_string());
      break;
    case rclcpp::ParameterType::PARAMETER_BOOL_ARRAY:
      blackboard->set<std::vector<bool>>(param_name, parameter.as_bool_array());
      break;
    case rclcpp::ParameterType::PARAMETER_INTEGER_ARRAY:
      blackboard->set<std::vector<int64_t>>(param_name,
                                            parameter.as_integer_array());
      break;
    case rclcpp::ParameterType::PARAMETER_DOUBLE_ARRAY:
      blackboard->set<std::vector<double>>(param_name,
                                           parameter.as_double_array());
      break;
    case rclcpp::ParameterType::PARAMETER_STRING_ARRAY:
      blackboard->set<std::vector<std::string>>(param_name,
                                                parameter.as_string_array());
      break;
    case rclcpp::ParameterType::PARAMETER_BYTE_ARRAY:
      blackboard->set<std::vector<uint8_t>>(param_name,
                                            parameter.as_byte_array());
      break;
    default:
      YASMIN_LOG_ERROR("Unsupported parameter type for '%s'",
                       param_name.c_str());
      return basic_outcomes::ABORT;
    }
  }

  return basic_outcomes::SUCCEED;
}

File: yasmin_ros/yasmin_ros/get_parameters_state.py
# Copyright (C) 2025 Miguel Ángel González Santamarta
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, Any

import rclpy
from rclpy.node import Node

import yasmin
from yasmin import State, Blackboard
from yasmin_ros.basic_outcomes import SUCCEED, ABORT, CANCEL
from yasmin_ros.ros_state_utils import resolve_node


class GetParametersState(State):
    """
    State that retrieves parameters from the ROS 2 parameter server.

    This state retrieves parameters from the ROS 2 parameter server and stores
    them in the blackboard.
    """

    def __init__(
        self,
        parameters: Dict[str, Any],
        node: Node = None,
    ) -> None:
        """
        Constructs a GetParametersState with a map of parameters.

        Args:
            parameters (Dict[str, Any]): A map of parameter names to their default values.
            node (Node, optional): A shared pointer to the ROS 2 node.
        """
        self._parameters = parameters
        self._node = resolve_node(node)

        super().__init__([SUCCEED, ABORT])

    def execute(self, blackboard: Blackboard) -> str:
        """
        Executes the state to retrieve parameters.

        Args:
            blackboard (Blackboard): A reference to the Yasmin blackboard.

        Returns:
            str: A string representing the outcome of the execution.
        """

        for param_name, param_value in self._parameters.items():
            if self.is_canceled():
                return CANCEL

            if not self._node.has_parameter(param_name):
                self._node.declare_parameter(param_name, param_value)

            yasmin.YASMIN_LOG_INFO(f"Retrieving parameter '{param_name}'")

            parameter = self._node.get_parameter(param_name)
            parameter_type = rclpy.Parameter.Type(
                self._node.get_parameter_type(param_name)
            )
            parameter_value = parameter.get_parameter_value()

            if parameter_type == rclpy.Parameter.Type.BOOL:
                value = parameter_value.bool_value
            elif parameter_type == rclpy.Parameter.Type.INTEGER:
                value = parameter_value.integer_value
            elif parameter_type == rclpy.Parameter.Type.DOUBLE:
                value = parameter_value.double_value
            elif parameter_type == rclpy.Parameter.Type.STRING:
                value = parameter_value.string_value
            elif parameter_type == rclpy.Parameter.Type.BOOL_ARRAY:
                value = parameter_value.bool_array_value
            elif parameter_type == rclpy.Parameter.Type.INTEGER_ARRAY:
                value = parameter_value.integer_array_value
            elif parameter_type == rclpy.Parameter.Type.DOUBLE_ARRAY:
                value = parameter_value.double_array_value
            elif parameter_type == rclpy.Parameter.Type.STRING_ARRAY:
                value = parameter_value.string_array_value
            elif parameter_type == rclpy.Parameter.Type.BYTE_ARRAY:
                value = parameter_value.byte_array_value
            else:
                yasmin.YASMIN_LOG_ERROR(
                    f"Unsupported parameter type for '{param_name}': {self._node.get_parameter_type(param_name)}"
                )
                return ABORT

            blackboard[param_name] = value

        return SUCCEED
